### ISSUE TITLE
NullReferenceException in CSharp target

### ISSUE THREAD

**Issue by slt3** [antlr/antlr4#2304]
I have following grammar:
```
grammar FormulaCalc;
expr: formula EOF;
formula 
	 : formula '^' formula 
	 | formula '*' formula
	 | formula '/' formula
	 | formula '%' formula
	 | formula '+' formula
	 | variable
	 | number
	 | formula '-' formula
     | '('formula')'
	 | function
     ;
function : zero_param_function
		 |one_param_function
		 |two_param_function;
zero_param_function : FUNNAME0 '('')';
one_param_function : FUNNAME1 '('parameter')';
two_param_function : FUNNAME2 '('parameter ',' parameter')';
parameter: number|variable;
FUNNAME0 : 'rand';
FUNNAME1 : 'sin'|'cos'|'tan'|'cot'|'asin'
		 |'acos'|'atan'|'acot'|'log'|'ln'|'exp';
FUNNAME2 :  'sqrt'|'pow'|'mod';
number : '-'? unsigned_number | '+'? unsigned_number;
unsigned_number : 'pi'| EULER |FPOINTNUMBER;
EULER : 'e'|'E';
FPOINTNUMBER : DIGIT+ ('.'DIGIT+)?;
DIGIT : [0-9];
variable : '{'word'}';
LETTER : [a-z]|[A-Z];
SPECIALLETTER : SIGN
	|'^'|'"'|'$'|'%'|'&'|'/'|'('|')'
	|'['|']'|'='|'?'|'#'|'_'|'.'|'~'
	|':'|','|';'|'<'|'>'
	;
word : (EULER|'^'|'*'|'/'|'%'|'+'|'-'|','
     |'('|')'
     |LETTER|FPOINTNUMBER|SPECIALLETTER
	 |FUNNAME0|FUNNAME1|FUNNAME2|'pi')+;
SIGN : '+'|'-';
WS : [ \t\r\n]+ -> skip;
```
When i run the parser:
```
String input = "5+1";
ICharStream stream = CharStreams.fromstring(input);
ITokenSource lexer = new FormulaCalcLexer(stream);
 ITokenStream tokens = new CommonTokenStream(lexer);
FormulaCalcParser parser = new FormulaCalcParser(tokens);
parser.BuildParseTree = true;
IParseTree tree = parser.expr();
```
the NullReferenceException is thrown:

Object reference not set to an instance of an object.
    Boolean Equals(System.Object)

   at Antlr4.Runtime.Atn.SingletonPredictionContext.Equals(Object o)
   at Antlr4.Runtime.Atn.ATNConfig.Equals(ATNConfig other)
   at Antlr4.Runtime.Atn.ATNConfig.Equals(Object o)
   at System.Collections.Generic.ObjectEqualityComparer`1.Equals(T x, T y)
   at System.Collections.Generic.HashSet`1.AddIfNotPresent(T value)
   at System.Collections.Generic.HashSet`1.Add(T item)
...

I tested it also with Java output, and it was working. So i checked the SingletonPredictionContext.Equals(Object o) implementation in Java and CSharp and they are slightly different. But if i make the code equal ( and check the parent for null), then it works also in CSharp

**Comment by ericvergnaud** [comment]
Hi
Thanks for this
Can you submit a PR?

Envoyé de mon iPhone

**Comment by slt3** [comment]
Sure i can. Wat was again a PR?

**Comment by ericvergnaud** [comment]
A pull request

Envoyé de mon iPhone

**Comment by KvanTTT** [comment]
Should it be closed?

**Comment by slt3** [comment]
Ok,
thanks

Von: ericvergnaud <notifications@github.com>
Gesendet: Freitag, 8. Januar 2021 13:36
An: antlr/antlr4 <antlr4@noreply.github.com>
Cc: Sebastian Linkiewicz <Sebastian.Linkiewicz@TSE-Systems.com>; Author <author@noreply.github.com>
Betreff: Re: [antlr/antlr4] NullReferenceException in CSharp target (#2304)


Closed #2304<https://github.com/antlr/antlr4/issues/2304>.

-
